Added isValidBubbleElevatorRelay, getBubbleElevatorDirection and convertIntoBubbleElevator  to IForgeBlock/IForgeBlockState

oOMitchOo opened the pull request:
~~Added the interface **IBubbleSource**, implemented it by **SoulSandBlock** and **MagmaBlock**. Then made **BubbleColumnBlock** check for instances of said interface instead of hardcoded vanilla Soul Sand or Magma Block.~~

Added _isValidBubbleElevatorRelay_ (default false for every Block not SoulSandBlock, MagmaBlock or BubbleColumnBlock), _getBubbleElevatorDirection_ (UP for SoulSandBlock, DOWN for MagmaBlock, UP/DOWN dependant on state for BubbleColumnBlock) and _convertIntoBubbleElevator_ (points to BubbleColumnBlock for vanilla water block) to **IForgeBlock** and **IForgeBlockState**

Patched **BubbleColumnBlock** to use _getBubbleElevatorDirection_ instead of its _getDrag_ method, check for _isValidBubbleColumnOrigin_ instead of hardcoded vanilla Soul Sand, Magma Block or Bubble Column in its _isValidPosition_ method and use _convertIntoBubbleElevator_ instead of its _placeBubbleColumn_ method.
Patched **SoulSandBlock** and **MagmaBlock** to use _convertIntoBubbleElevator_.

This way it is possible for modders to add a block, which (spawns and) maintains a vanilla bubble elevator in vanilla water. And to add their own bubble elevator blocks, which can mix with vanilla bubble column blocks.

I chose to not go the block tag route, since tagging random blocks wouldn't be enough to make them a bubble elevator source. The block would always need to call placeBubbleColumn in its tick method.
This was requested in #5458.

I tested it and it works great.

oOMitchOo committed:
Added the interface IBubbleSource, implemented it by SoulSandBlock and MagmaBlock. Then made BubbleColumnBlock check for instances of said interface instead of hardcoded vanilla Soul Sand or Magma Block. This way it is possible for modders to add a block, which spawns and maintains a bubble elevator in vanilla water. I chose to not get the block tag route, since tagging random blocks wouldn't be enough to make them a bubble elevator source. The block would always need to call placeBubbleColumn in its tick method.

williewillus commented on patches/minecraft/net/minecraft/block/IBubbleSource.java.patch:
```diff
@@ -0,0 +1,26 @@
++    }
++
++    boolean isDragging();
++}
```
two things:
1. why is this in the patches folder instead of forge sources?
2. You can't introduce methods "above" the vanilla class hierarchy with the same name like your do here since the vanilla methods will be reobfed but these won't.

oOMitchOo replied:
1. Right. I have to change that.
2. I see. So basically drop every vanilla method in the interface? It wouldn't result in any errors if they're missing anyway. A functioning bubble source block would still need them.

oOMitchOo committed:
Put the interface in the right place (forge) and simplified it.

williewillus commented on src/main/java/net/minecraftforge/block/IBubbleSource.java:
```diff
@@ -0,0 +1,16 @@
+     * This should return TRUE, if block is maintaining a bubble column block with downward drag. (like Magma Block)
+     * And it should return FALSE, if it is maintaining a bubble column block with upward pushing. (like Soul Sand)
+     */
+    boolean isDragging();
```
you could add a default impl here for soulsand/bubble blocks which would remove the need to patch an isDragging override into those two blocks

oOMitchOo replied:
Good idea!

oOMitchOo committed:
Added default isDragging method for IBubbleSource. Less patching of vanilla SoulSandBlock and MagmaBlock.

pupnewfster commented on patches/minecraft/net/minecraft/block/BubbleColumnBlock.java.patch:
```diff
@@ -0,0 +1,33 @@
+-         return block != Blocks.field_150425_aM;
+       }
++      if (block instanceof IBubbleSource) {
++         return ((IBubbleSource) block).isDragging();
```
From my understanding of modifying vanilla code (that creates the patches) we are not supposed to add more imports so as to reduce the patch size. So that would make this become:
```java
if (block instanceof net.minecraftforge.block.IBubbleSource) {
    return ((net.minecraftforge.block.IBubbleSource) block).isDragging();
```
Realistically to further reduce the patch size it potentially should be (keeping the else that already exists):
```java
return !(block instanceof net.minecraftforge.block.IBubbleSource) || ((net.minecraftforge.block.IBubbleSource) block).isDragging();
```

pupnewfster commented on patches/minecraft/net/minecraft/block/MagmaBlock.java.patch:
```diff
@@ -0,0 +1,13 @@
++import net.minecraftforge.block.IBubbleSource;
+ 
+-public class MagmaBlock extends Block {
++public class MagmaBlock extends Block implements IBubbleSource {
```
And this would become:
```java
public class MagmaBlock extends Block implements net.minecraftforge.block.IBubbleSource {
```

pupnewfster commented on patches/minecraft/net/minecraft/block/SoulSandBlock.java.patch:
```diff
@@ -0,0 +1,13 @@
++import net.minecraftforge.block.IBubbleSource;
+ 
+-public class SoulSandBlock extends Block {
++public class SoulSandBlock extends Block implements IBubbleSource {
```
And this would become:
```java
public class SoulSandBlock extends Block implements net.minecraftforge.block.IBubbleSource {
```

diesieben07 reviewed (changes requested):
Usually Forge does not use interfaces like `IBubbleSource`. Instead add a method in `IForgeBlock`, which is implemented by all blocks.

oOMitchOo commented:
Ah, I see. I guess I should try the IForgeBlock route then.

oOMitchOo committed:
Added isBubbleSource and getDrag in IForgeBlock instead of using an interface and patching vanilla Soul Sand and Magma Block.

This way the classes SoulSandBlock and MagmaBlock won't get patched and the BubbleColumnBlock patch is as close to the vanilla version as possible.

I've tested it with a block, which extends SoulSandBlock and overrides isBubbleSource to return true and getDrag to return false.
I didn't put the test into the pull request since I don't think this needs further testing.

LexManos commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:
```diff
@@ -1037,4 +1037,23 @@ default boolean collisionExtendsVertically(BlockState state, IBlockReader world,
+     * NOTE: This will return true for your Block by default, but it is never called, as long as a BubbleColumnBlock
+     * isn't added above your block and the BubbleColumnBlock won't stay, if IForgeBlock.isBubbleSource returns false.
+     */
+    default boolean getDrag() {
```
getDrag is a bad name, it is to ambiguous.

LexManos commented on patches/minecraft/net/minecraft/block/BubbleColumnBlock.java.patch:
```diff
@@ -0,0 +1,20 @@
+    public boolean func_196260_a(BlockState p_196260_1_, IWorldReader p_196260_2_, BlockPos p_196260_3_) {
+       Block block = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b()).func_177230_c();
+-      return block == Blocks.field_203203_C || block == Blocks.field_196814_hQ || block == Blocks.field_150425_aM;
++      return block == Blocks.field_203203_C || block.isBubbleSource();
```
This doesn't support custom bubble column blocks, as in blocks that act as a bubble relay, but not a source.

LexManos reviewed (changes requested):
This needs to be expanded into a fully thought out system.
There are two side of Bubble columns. The source and the bubbles themselves.
Would this be better as a integer value, instead of a simple boolean?
Such as allowing a 'weak' sources that only travel 10 blocks?
Its not something we could add onto the vanilla Bubble Column directly, so having them have a'power' of 256 would make sense.
But having it a int would allow someone else to come along and add a new bubble block that supports power.

oOMitchOo commented:
I mean, **modders can always add their own bubble column block** (for vanilla water) and make their bubble source spawn that one instead of the vanilla one.

And then the possibilities are endless. They would have full control over the creation and maintenance of those bubble columns.

So, this would only be useful as a kind of reuse purpose or to **ensure better compatibility between mods and vanilla MC**. Plus you won't have 10 Mods, which create 10 different bubble column blocks in vanilla water. (10 different blocks, which basically work the same way, that is)

That being said. I agree, that the method names aren't perfect. I'm not native english speaking, if that's an excuse. Maybe **isBubbleSource could be** something like **canMaintainBubbleColumn** and I don't know about getDrag... nothing I can think of sounds good.

TeamSpen210 commented:
`producesBubbleColumn()` and `isBubbleColumnDragger()`? For custom bubble columns, I think the idea would be to allow said custom blocks to support the original columns on top of them. That could be resolved though by passing in blockstate and world - which would probably be good to do anyway.

oOMitchOo committed:
Clearer method names and descriptions.

Title says it all. Plus I've used the Direction Property, which already
was imported in both IForgeBlock and BubbleColumnBlock to make it
clearer in which direction the resulting bubble elevator will push you.

My opinion is, that any modder, who wants a more complicated behavior
of the bubble column block, can add a modded version and create that
one with his 'bubble source block' instead of the vanilla one.

oOMitchOo commented:
Regarding the idea with the int value: Modders can always add a modded version of BubbleColumnBlock and spawn that one instead of the vanilla version. So I don't think this needs to be done by Forge.

This commit basically wants to give modders, who don't want to add their own bubble block, the possibility to create the vanilla version instead.

LexManos commented:
I can see what this PR is intending to do. My point is that it's only 1/2 the equation.
The other half, custom bubbles, needs to be addressed.
You don't want the spawner, to know about the specific spawned block.
You just want to have things like soul sand to ask the block on top of it to convert to a bubble.
That way say... if someone wants to come along and make trap doors that can be both waterlogged, and bubbled, so that when closed they get stuck, and when open they fly through it, they can.
All Soul Sand needs to know is that the block above it is water logged, and then ask it to turn into a bubble column.

oOMitchOo commented:
OK, I see what you mean.
~~It would be fitting to put that into ILiquidContainer, though I guess the preferred way to go about this is again to add methods to IForgeBlock because less patching of vanilla classes?~~

oOMitchOo committed:
Made everything BlockState dependant

Plus, I've added a test mod with a trapdoor/bubbleColumn hybrid.

It works great as far, as I can see. But I need to do some further
testing. There's double dipping under water sound for example.
Possibly just a bug in my test mod.

oOMitchOo commented:
Sorry for bothering you @diesieben07. I'm not 100% sure about the last commit and wanted another opinion before bothering LexManos.

oOMitchOo commented:
SoulSandBlock and MagmaBlock need patching.

Right now everything works fine as long as the first block above soul sand or magma block is a water source block.

oOMitchOo committed:
Little better naming & description of methods

Every bubble elevator related method now has bubble elevator in its
name. Plus I've changed their descriptions a little.

An unnecessary check in BubbleColumnBlock#isValidPosition was removed.

oOMitchOo commented:
OK, I personally like this solution quite well. **Custom bubble elevator blocks in between vanilla bubble column blocks are now possible.** The creation of a bubble elevator is still very vanilla-like in that it's like a chain reaction beginning from the soul sand or magma block.

The only thing that bothers me and maybe needs discussing is, that **a vanilla water block will always turn into a BubbleColumnBlock**. Which could not be wanted by modders, since the vanilla bubble column is pretty limited in its functionality.

I could change the convertIntoBubbleElevator method to also take a **Nullable Block, which replaces the vanilla bubble column**, if not null. But this would only work with modded versions of Soul Sand and Magma Block...

Alpvax commented:
An alternative approach would be to fire a BubbleColumnCreate event, which modders could use to override the bubble column block.

oOMitchOo commented:
> An alternative approach would be to fire a BubbleColumnCreate event, which modders could use to override the bubble column block.

But wouldn't that be a problem, if **two different mods** wanted to create a new bubble column block, which would be created in the place of a vanilla water block?

I mean, the only problem I see with putting that nullable Block (as a replacement for BubbleColumnBlock) in the method, would be, that **mixing blocks** (which can be part of a bubble elevator) of different mods could end up in a bubble elevator containing **different BubbleColumnBlock replacements at different heights** of the column. Always dependant on the block, which came before the vanilla water block. Which would be a little confusing but natural?

Modders can always decide to write bubble elevator blocks, which don't mix with vanilla or other modded bubble elevator blocks.

Alpvax commented:
> But wouldn't that be a problem, if two different mods wanted to create a new bubble column block, which would be created in the place of a vanilla water block?

The last event would be the winner, as with all other events.

I'm not sure how your approach fixes the issue, it just means nothing else can override the placement. If you really wish to do that, place the column block manually, but it sounds like a bad idea to me (what happens for a custom fluid which the column generator doesn't know about?).

Alpvax commented:
> mixing blocks (which can be part of a bubble elevator) of different mods could end up in a bubble elevator containing different BubbleColumnBlock replacements at different heights of the column. Always dependant on the block, which came before the vanilla water block. Which would be a little confusing but natural?

I don't think the original block matters for higher up blocks. Once the initial block has been replaced, all other mod blocks are dependant on the block directly below them.
E.g. mod A swaps the direction of soul sand (using the event). Mod B's fluid should then only care about the direction of the column below it, otherwise the column changes direction part way through.

oOMitchOo commented:
I mean this scenario... let's say mod A adds a custom bubble column (ABub) and a custom waterloggable block (ABlo), which also can be part of a bubble elevator. Mod B also adds a custom bubble column (BBub) and a waterloggable block (BBlo).

ABlo and BBlo can be in elevator mode (+: pushing entities) and in normal waterlogged mode (-: not pushing). And ABlo will convert vanilla Water into ABub, while BBlo will convert it into BBub.

We have this water column. (vaBub is a vanilla BubbleColumnBlock)
**First arrow** is after adding some modded waterloggable Blocks, **second arrow** is after adding Soul Sand on the bottom and after the whole chain reaction creating a bubble elevator by calling convertIntoBubbleElevator. This would be happening when adding a nullable block parameter into the method.
<pre>
+-------+            +-------+            +-------+
| Water |            | Water |            | BBub  |
+-------+            +-------+            +-------+
| Water |            | Water |            | BBub  |
+-------+            +-------+            +-------+
| Water |            | BBlo- |            | BBlo+ |
+-------+     ->     +-------+     ->     +-------+
| Water |            | Water |            | ABub  |
+-------+            +-------+            +-------+
| Water |            | ABlo- |            | ABlo+ |
+-------+            +-------+            +-------+
| Water |            | Water |            | vaBub |
+-------+            +-------+            +-------+
| Sand  |            | Sand  |            | SoulS |
+-------+            +-------+            +-------+
</pre>
That should be acceptable right? Modders can always add some checks before calling convertIntoBubbleElevator to make sure their bubble elevator blocks don't mix with those of other mods.

Alpvax commented:
> **First arrow** is after adding some modded waterloggable Blocks, **second arrow** is after adding Soul Sand on the bottom and after the whole chain reaction creating a bubble elevator by calling convertIntoBubbleElevator. This would be happening when adding a nullable block parameter into the method.
> 
> That should be acceptable right? Modders can always add some checks before calling convertIntoBubbleElevator to make sure their bubble elevator blocks don't mix with those of other mods.

That would work the same with an event. And BBlo doesn't need to know anything about the soul sand, just that the block below is a "pushing" bubble column.
In this scenario, why would A and B need their own implementation of a bubble block? Why could they not just use the vanilla one (and only have a custom implementation of the waterlogged block)?
What happens with other fluids? Say ModC adds a block which creates bubbles in lava. Now ABub and BBub don't apply, because they are bubbles in water.

I understand that currently fluid support isn't complete, but I think we still need to consider it to save completely reworking things in future.
Lex's approach of asking the block to convert is a better approach, let the fluid decide what the bubble block is, rather than the source. The advantage of an event is that that decision can be overridden in the rare cases where we would need to.

oOMitchOo commented:
> And BBlo doesn't need to know anything about the soul sand, just that the block below is a "pushing" bubble column.

I know. That's how it works right now. It's like a chain reaction. Soul Sand calling convertIntoBubbleElevator on the block above, then that one calling convertIntoBubbleElevator on the next block and so on. Just like in vanilla. (if modders do it correctly that is)

> In this scenario, why would A and B need their own implementation of a bubble block? Why could they not just use the vanilla one (and only have a custom implementation of the waterlogged block)?

They would not need a custom bubble block. They would only need a custom bubble block, if they wanted some more complex behavior like for example stopping the creation of new elevator blocks at a certain y-level. I only wanted to make the case, since mixing modded custom bubble columns also means mixing custom behavior. (using the current example: the elevator would possibly not stop at a certain y-level)

> What happens with other fluids? Say ModC adds a block which creates bubbles in lava. Now ABub and BBub don't apply, because they are bubbles in water.

Since Soul Sand calls convertIntoBubbleElevator on the block above it, it can also work with other fluids, when those other fluid's blocks are coded accordingly. So it wouldn't work with lava, since lava only has the default convertIntoBubbleElevator method. But it would work with every modded fluid, if the modder wants it.

> I understand that currently fluid support isn't complete, but I think we still need to consider it to save completely reworking things in future.

I don't completely understand what you're saying here. I know, that there were talks about enhancing the fluid system in MC a lot. But my opinion is, that it could still take a lot of time until that'll be done and until then it would be great to focus on the compatibility of vanilla objects with modded objects.

> Lex's approach of asking the block to convert is a better approach, let the fluid decide what the bubble block is, rather than the source. The advantage of an event is that that decision can be overridden in the rare cases where we would need to.

But that's exactly what's happening at the moment. Soul Sand is calling convertIntoBubbleElevator on the block above. If that block has overriden the method, it will create its corresponding elevator block (or block state). In my example Soul Sand would for example make ABlo- convert itself into ABlo+.

The only problem is, when using convertIntoBubbleElevator on a vanilla water block. It will always convert itself into a vanilla bubble elevator, which would sometimes not be desirable, because the vanilla bubble column block has a limited functionality. That's why I discussed adding a nullable parameter to the method.

I think we're talking past each other a little...

oOMitchOo commented:
My test mod acts exactly like the example Lex gave.
I've added a trapdoor, which can also be part of a bubble elevator. And the Soul Sand / Magma Block / BubbleColumnBlock doesn't need to know about the trapdoor. It is just calling convertIntoBubbleElevator on the block above. If that block is the trapdoor, it will change its state.

oOMitchOo committed:
Added a Nullable Block parameter to convertIntoBubbleElevator

Calling convertIntoBubbleElevator on a vanilla water block would always
convert it into a vanilla bubble column block.

With this Nullable Block parameter it will only do so, if the parameter
is null otherwise it will use the non-null Block's
convertIntoBubbleElevator method.

That way modders can not only add custom waterloggable blocks, which can
be part of the bubble elevator but also make sure vanilla water blocks
are convert into modded bubble column blocks, which can have new
functionality. (for example stopping the creation of new bubble elevator
blocks at a certain y-level) It should be noted though that mixing
different modded blocks in a water elevator can also result in mixing
functionality. Modders need to think about more checks, if they don't
want this to happen.

Alpvax commented:
I think there is an element of us talking past each other, yes.

However, my main argument is for an event rather than the vanillaBubbleReplacement argument. It would allow things such as forcing ALL water bubble columns to be restricted to a certain height. (You could pass in the previous height, and keep incrementing for each similar block). There is nothing to stop modders manually placing a custom bubble column block (and doing the same checks that you suggested before).

I do believe that we will not resolve this discussion ourselves, so I will remain quiet until someone else weighs in with an opinion.

LexManos commented on patches/minecraft/net/minecraft/block/SoulSandBlock.java.patch:
```diff
@@ -0,0 +1,19 @@
+ import net.minecraft.entity.Entity;
+ import net.minecraft.entity.EntityType;
+ import net.minecraft.pathfinding.PathType;
++import net.minecraft.util.Direction;
```
Import

LexManos commented on patches/minecraft/net/minecraft/block/BubbleColumnBlock.java.patch:
```diff
@@ -0,0 +1,33 @@
++      } else {
++         p_203159_0_.func_180495_p(p_203159_1_).convertIntoBubbleElevator(p_203159_0_, p_203159_1_, null, p_203159_2_ ? Direction.DOWN : Direction.UP);
+       }
+-
```
Spam

LexManos commented on patches/minecraft/net/minecraft/block/BubbleColumnBlock.java.patch:
```diff
@@ -0,0 +1,33 @@
+-      Block block = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b()).func_177230_c();
+-      return block == Blocks.field_203203_C || block == Blocks.field_196814_hQ || block == Blocks.field_150425_aM;
++      BlockState blockStateBelow = p_196260_2_.func_180495_p(p_196260_3_.func_177977_b());
++      return blockStateBelow.isValidBubbleElevatorRelay();
```
Single line

LexManos commented on patches/minecraft/net/minecraft/block/BubbleColumnBlock.java.patch:
```diff
@@ -0,0 +1,33 @@
+-
+    }
+ 
+    public static boolean func_208072_b(IWorld p_208072_0_, BlockPos p_208072_1_) {
```
Needs @Deprecated and describing what to use instead.

LexManos commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlock.java:
```diff
@@ -1037,4 +1027,59 @@ default boolean collisionExtendsVertically(BlockState state, IBlockReader world,
+     *                                       parameter being null again converting vanilla water blocks into vanilla
+     *                                       bubble column blocks.
+     */
+    default void convertIntoBubbleElevator(BlockState state, IWorld world, BlockPos pos, @Nullable Block vanillaBubbleColumnReplacement, Direction elevatorDirection) {
```
Why is it a Block? BlockState is the new block!

LexManos commented on src/main/java/net/minecraftforge/common/extensions/IForgeBlockState.java:
```diff
@@ -879,4 +879,30 @@ default boolean collisionExtendsVertically(IBlockReader world, BlockPos pos, Ent
+    default Direction getBubbleElevatorDirection() {
+        Block block = this.getBlockState().getBlock();
+        if (block == Blocks.BUBBLE_COLUMN) {
+            return this.getBlockState().get(BubbleColumnBlock.DRAG) ? Direction.DOWN : Direction.UP;
```
This logic should be in the default implementation.

oOMitchOo committed:
Applied all change requests

For example is the Nullable replacement for a vanilla water column now
a BlockState instead of a Block.

BubbleColumnBlock#getDrag isn't used anymore and marked as Deprecated.

Every method in IForgeBlockState now directly calls the corresponding
method in IForgeBlock, which then does all the work.
